src/Tools/dotnet-openapi/src/Program.cs

```diff
@@ -0,0 +1,441 @@
+        {
+            return c.Argument(SourceFileArgName, $"The openapi file to {action}. This can be a path to a local openapi file, " +
+                "a URI to a remote openapi file or a path to a *.csproj file containing openapi endpoints");
+        }
+
+        private FileInfo ResolveProjectFile(CommandOption optProjects)
+        {
+            string csproj;
+            if(optProjects.HasValue())
+            {
+                csproj = optProjects.Value();
+                csproj = Path.Combine(_workingDir, csproj);
+                if(!File.Exists(csproj))
+                {
+                    _reporter.Error("The given csproj does not exist.");
+                }
+            }
+            else
+            {
+                var csprojs = Directory.GetFiles(_workingDir, "*.csproj", SearchOption.TopDirectoryOnly);
+                if(csprojs.Length == 0)
+                {
+                    _reporter.Error("No csproj files were found in the current directory. Either move to a new directory or provide the project explicitly");
+                }
+                if(csprojs.Length > 1)
+                {
+                    _reporter.Error("More than one csproj was found in this directory, either remove a duplicate or explicitly provide the project.");
+                }
+
+                csproj = csprojs.Single();
+            }
+
+            return new FileInfo(csproj);
+        }
+
+        private void EnsurePackagesInProject(FileInfo projectFile, CodeGenerator codeGenerator)
+        {
+            var packages = GetServicePackages(codeGenerator);
+            foreach (var (packageId, version) in packages)
+            {
+                var args = new string[] {
+                    "add",
+                    "package",
+                    packageId,
+                    "--version",
+                    version
+                };
+
+                var startInfo = new ProcessStartInfo
+                {
+                    FileName = DotNetMuxer.MuxerPath,
+                    Arguments = string.Join(" ", args),
+                    WorkingDirectory = projectFile.Directory.FullName,
+                    RedirectStandardError = true,
+                    RedirectStandardOutput = true,
+                };
+
+                var tcs = new TaskCompletionSource<int>();
+
+                var process = Process.Start(startInfo);
+                process.WaitForExit(20 * 1000);
+
+                if (process.ExitCode != 0)
+                {
+                    _reporter.Error(process.StandardError.ReadToEnd());
+                    _reporter.Error($"Could not add package `{packageId}` to `{projectFile.Directory}`");
+                    throw new ArgumentException();
+                }
+            }
+        }
+
+        private XmlDocument LoadProject(FileInfo projectFile, out XmlElement projNode)
+        {
+            var projXml = new XmlDocument();
```

Comment:
MSBuild project evaluation is more nuanced than parsing XML. I don’t know the requirements of this tool, but if projects are allowed to define conditional items, vary things by target framework, or use Imports, the simple XML evaluation is going to be insufficient. We can chat about the right way to evaluate and edit an MSBuild file if that’s the case.